lib/core/utils/preload.js

```diff
@@ -0,0 +1,136 @@
+	return (
+		typeof preload === 'object' &&
+		preload.hasOwnProperty('assets') &&
+		Array.isArray(preload.assets) &&
+		preload.assets.length
+	);
+}
+
+/**
+ * Returns a boolean which decides if preload is configured
+ * @param {Object} options run configuration options (or defaults) passed via axe.run
+ * @return {boolean}
+ */
+function shouldPreload(options) {
+	if (!options) {
+		return false;
+	}
+
+	if (!options.preload) {
+		return false;
+	}
+
+	if (typeof options.preload === typeof true) {
+		return options.preload;
+	}
+
+	if (isPreloadValidObject(options.preload)) {
+		return true;
+	}
+
+	return false;
+}
+axe.utils.shouldPreload = shouldPreload;
+
+/**
+ * Constructs a configuration object representing the preload requested assets & timeout
+ * @param {Object} options run configuration options (or defaults) passed via axe.run
+ * @return {Object}
+ */
+function getPreloadConfig(options) {
+	// default fallback configuration
+	let out = {
+		assets: axe.constants.preloadAssets,
+		timeout: axe.constants.preloadAssetsTimeout
+	};
+
+	// if type is boolean
+	if (typeof options.preload === typeof true) {
+		return out;
+	}
+
+	// if type is object - ensure an array of assets to load is specified
+	if (isPreloadValidObject(options.preload)) {
+		const requestedAssets = [];
+		options.preload.assets.forEach(asset => {
+			const a = asset.toLowerCase();
+			if (axe.constants.preloadAssets.includes(a)) {
+				// unique assets to load, in case user had requested same asset type many times.
+				if (!requestedAssets.includes(a)) {
+					requestedAssets.push(a);
+				}
+				return a;
+			} else {
+				const e =
+					`Requested asset: ${a}, not supported by aXe.` +
+					`Supported assets are: ${axe.constants.preloadAssets
+						.map(_ => _)
+						.join(', ')}.`;
+				throw new Error(e);
+			}
+		});
+		out.assets = requestedAssets;
+
+		if (options.preload.timeout) {
+			if (
+				typeof options.preload.timeout === 'number' &&
+				!Number.isNaN(options.preload.timeout)
+			) {
+				out.timeout = options.preload.timeout;
+			} else {
+				throw new Error(`preload timeout specified is not of type number`);
+			}
+		}
+
+		return out;
+	} else {
+		throw new Error(
+			'No assets configured for preload in aXe run configuration'
+		);
+	}
+}
+axe.utils.getPreloadConfig = getPreloadConfig;
+
+/**
+ * Returns a then(able) queue with results of all requested preload(able) assets. Eg: ['cssom'].
+ * If preload is set to false, returns an empty queue.
+ * @param {Object} options run configuration options (or defaults) passed via axe.run
+ * @return {Object} queue
+ */
+function preload(options) {
+	const preloadFunctionsMap = {
+		cssom: axe.utils.preloadCssom
+	};
+
+	const q = axe.utils.queue();
+
+	const shouldPreload = axe.utils.shouldPreload(options);
+	if (!shouldPreload) {
+		return q;
+	}
+
+	const preloadConfig = axe.utils.getPreloadConfig(options);
+
+	preloadConfig.assets.forEach(asset => {
+		q.defer((resolve, reject) => {
```

Comment:
What is the purpose of this top-level `.defer()`? It looks like this could be rewritten as:

```
prelaodedConfig.assets.forEach(asset => {
  preloadFunctionsMap[asset]({ asset, timeout: preloadConfig.timeout })
    .then(results => results[0])
})
```

Reply:
prefer to keep the top level `q`, then all the preloaded asset types lives under one `q`, when is then(ed) in the `run` setup.

Reply:
The queue isn't chainable, so that wouldn't work. It only gives you one "than".

Reply:
Oh, weird. I figured it worked the same way as a native Promise.

Reply:
Still keeping this one, as is.